{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f533a539-1898-4231-af15-62174d1c1ae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "from collections import Counter\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import (MultipleLocator, AutoMinorLocator)\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import re\n",
    "import sys\n",
    "from scipy import stats\n",
    "from scipy.stats import mannwhitneyu\n",
    "import matplotlib as mpl\n",
    "import functools "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6154428-6075-409a-a7ff-c6a9a370ec63",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "## /lab/users/jhonDoe/Code/Get_cover_rust_pysam\n",
    "sys.path.append(\"<path to coverage_rust_bam directory>\")\n",
    "## this is import the  python lib that uses the brust backend \n",
    "## see readme for install \n",
    "import Get_cover "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5ef07f9-1395-4516-b871-7bac7bec834e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attr(string):\n",
    "    dico = {}\n",
    "    spt_ = string.split(\";\")\n",
    "    \n",
    "    for x in spt_:\n",
    "        if x:\n",
    "            try:\n",
    "                dico[x.split()[0].strip()] =  x.split()[1].replace('\"', \"\").strip()\n",
    "            except:\n",
    "                print(spt_)\n",
    "                raise\n",
    "    return dico\n",
    "\n",
    "\n",
    "def gtf_to_dict(gtf_file, main_ = \"gene_id\"):\n",
    "    print(main_)\n",
    "    dico = {}\n",
    "    with open(gtf_file) as f_in:\n",
    "        \"\"\"atm reads only genes\"\"\"\n",
    "        for line in f_in:\n",
    "            line = line.strip()\n",
    "            if not line:\n",
    "                continue\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            spt = line.strip().split(\"\\t\")\n",
    "            try:\n",
    "                chr_ = spt[0]\n",
    "                start = int(spt[3]) \n",
    "                end = int(spt[4]) \n",
    "                strand =  spt[6]\n",
    "                attr = get_attr(spt[-1])\n",
    "                gene_id = attr[main_]\n",
    "                gene_symbol = attr.get(\"gene_symbol\", \"None\")\n",
    "                type_ = spt[2]\n",
    "            except:\n",
    "                print(attr)\n",
    "                print(line)\n",
    "                print(spt)\n",
    "                raise\n",
    "            \n",
    "            if gene_id not in dico:\n",
    "                dico[gene_id] = {\n",
    "                    \"chr\": chr_,\n",
    "                    \"symbol\" :  gene_symbol,\n",
    "                    \"strand\" : strand,\n",
    "                    \"transcript\": {}\n",
    "                }\n",
    "                \n",
    "            if type_ == \"gene\":\n",
    "    \n",
    "                dico[gene_id][\"start\"] = start\n",
    "                dico[gene_id][\"end\"] = end\n",
    "            \n",
    "            else:\n",
    "                try:\n",
    "                    transcript_id  = attr.get(\"transcript_id\", \"None\")\n",
    "                    transcript_symbol = attr.get(\"transcript_symbol\", \"None\")\n",
    "\n",
    "                    \n",
    "                    if transcript_id not in dico[gene_id][\"transcript\"]:\n",
    "                        dico[gene_id][\"transcript\"][transcript_id] = {\n",
    "                        \"transcript_symbol\" : transcript_symbol,\n",
    "                        \"transcript_id\" : transcript_id\n",
    "                        }\n",
    "                    \n",
    "                    if type_ not in dico[gene_id][\"transcript\"][transcript_id]: \n",
    "                        dico[gene_id][\"transcript\"][transcript_id][type_] = []\n",
    "                        \n",
    "                    dico[gene_id][\"transcript\"][transcript_id][type_].append({\n",
    "                        \"chr\": chr_,\n",
    "                        \"start\": start, \n",
    "                        \"end\" : end,\n",
    "                        \"strand\" : strand,\n",
    "                    })\n",
    "                except: \n",
    "                    continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaeeeec0-19c9-4cd4-a591-76226b3d1042",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = \"<path to annotated Y genes gtf>\"\n",
    "gtf_dict = gtf_to_dict(gtf_file, \"gene_id\") # if you don't use the gtf we provides be sure there is a \"gene_id\" attributes\n",
    "bam_dir = Path(\"<path to directory with bam files>\")\n",
    "# bam dir with alignment (of D.mauritiana and hybrid) against D. mauritiana inside all hybrids file start with trimmed_sim_mau\n",
    "# optionaly give a list of the path of all bam files in bam_files variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e7b588",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_file = list(bam_dir.glob(\"*bam\"))\n",
    "bam_file.sort()\n",
    "\n",
    "# check those bam_file are the one you want!\n",
    "for bam in bam_file:\n",
    "    print(bam)\n",
    "\n",
    "# if not copy paste the one you want \n",
    "# bam_file = [... ,... , ...]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13413212-6164-4f33-8ed6-6948c6a76bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#bam_file = [str(bam_dir /  x) for x in bams]\n",
    "#{bam_file.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dfebcd7-c2b8-44d2-a113-aaa24a4541f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir=\"<Path to outdir>\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a85ff6-9687-466d-a5bf-572c2a2d1ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in gtf_dict:\n",
    "\n",
    "    for transcript in gtf_dict[g][\"transcript\"]:\n",
    "        intervalls = gtf_dict[g][\"transcript\"][transcript][\"exon\"]\n",
    "\n",
    "        for e in intervalls:\n",
    "            e[\"type\"] = \"exon\"\n",
    "\n",
    "        results = Get_cover._get_Coverage_intervall(bam_file, intervalls=intervalls, n_thread=8, lib_scheme=\"frFirstStrand\", mapq=13, flag_in=0, flag_out=256)\n",
    "        print(\"pool done\")\n",
    "        for bam in results:\n",
    "            results[bam].max_reads = Get_cover.get_max_read(bam)\n",
    "            results[bam].normalize()\n",
    "        \n",
    "        \n",
    "        res = {}\n",
    "        for file in bam_file:\n",
    "            if \"L001\" in file:\n",
    "                f2 = file.replace(\"L001\", \"L002\")\n",
    "                X = Path(file).stem\n",
    "                res[X] = Get_cover.Coverage.merge_duplicates([results[file], results[f2]])\n",
    "                                                        \n",
    "                                                         \n",
    "        c2 = [\"#abd9e9\",\"#74add1\",\"#4575b4\",\"#313695\"] \n",
    "        c1 = [\"#d73027\",\"#f46d43\",\"#fdae61\",\"#fee090\"]\n",
    "\n",
    "        dico_colors = {}\n",
    "        c1_cpt = 0\n",
    "        c2_cpt = 0\n",
    "        for bam in res:\n",
    "            if \"trimmed_sim_mau\" in Path(bam).stem:\n",
    "                dico_colors[bam] = c1[c1_cpt]\n",
    "                c1_cpt += 1\n",
    "\n",
    "            else:\n",
    "                dico_colors[bam] = c2[c2_cpt]\n",
    "                c2_cpt += 1\n",
    "\n",
    "        dico_labels = dict((x, x) for x in  res)\n",
    "\n",
    "        Get_cover.plot(res, N=None, intron=\"exon\",\n",
    "                       intron_prop=0.35,\n",
    "                          dico_color=dico_colors,\n",
    "                          title=gtf_dict[g][\"symbol\"]+\" \"+g,\n",
    "                        out=out_dir + gtf_dict[g][\"symbol\"] + \"_exon_mau.pdf\",\n",
    "                       linewidth=3)\n",
    "        plt.show();"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
