{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a12190f-b3f5-4351-bb8c-fd10586d4328",
   "metadata": {},
   "outputs": [],
   "source": [
    "### import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.patches import Circle, Polygon, Wedge\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08db74bd-232d-4743-9215-080357034be9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Rust_alnPairMat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db7db53f-6fed-4bd6-8242-015b817380e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_seq_one_by_one(open_file, position=None):\n",
    "    \"\"\"\n",
    "\n",
    "    An Iterator over an opened fasta file\n",
    "    :param open_file: an open fasta file\n",
    "    :return: tuple(prompt, sequence)\n",
    "    \"\"\"\n",
    "\n",
    "    pos = 0\n",
    "    p, seq = \"\", \"\"\n",
    "    line = open_file.readline()\n",
    "    while line:\n",
    "\n",
    "        if line.startswith('>'):\n",
    "\n",
    "            if seq:\n",
    "\n",
    "                if not position:\n",
    "                    yield p, seq\n",
    "                else:\n",
    "                    yield p, seq, pos\n",
    "                p, seq = \"\", \"\"\n",
    "\n",
    "            p = line[1:].strip()\n",
    "\n",
    "        else:\n",
    "            seq += line.strip()\n",
    "            pos = open_file.tell()\n",
    "        line = open_file.readline()\n",
    "\n",
    "    if not position:\n",
    "        yield p, seq\n",
    "    else:\n",
    "        yield p, seq, pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddcdc5e4-c093-4f56-84f1-bd28267af8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(x):\n",
    "    d = { \"N\": \"N\",\n",
    "        \"A\": \"T\", \n",
    "         \"T\": \"A\",\n",
    "         \"C\": \"G\",\n",
    "         \"G\": \"C\"}\n",
    "    return ''.join([d[e] for e in x[::-1]])\n",
    "\n",
    "def get_intronseq(header, dico_seq):\n",
    "    \n",
    "\n",
    "    strand = header.split(\"__\")[-1]\n",
    "    #mau_exon_size = []\n",
    "    intron_res = []\n",
    "    contig = header.split(\"_FB\")[0]\n",
    "    exons = []\n",
    "\n",
    "    for r in header.split(\"__\")[1].split(\";\"):\n",
    "        r = r.split('...')\n",
    "        exons.append((int(r[0]), int(r[1])))\n",
    "    for i, r in enumerate(exons[0:-1]):\n",
    "        introns = dico_seq[contig][r[1]:exons[i+1][0]].upper()\n",
    "        if strand == \"-\":\n",
    "            introns = reverse_complement(introns)\n",
    "        intron_res.append(introns)\n",
    "    if strand == \"-\":\n",
    "        intron_res = intron_res[::-1]\n",
    "    return intron_res\n",
    "\n",
    "def clamp(v, minimum, maximum):\n",
    "    if v < minimum:\n",
    "        v = minimum\n",
    "    elif v > maximum:\n",
    "        v = maximum\n",
    "    return v    \n",
    "def get_windows_size(seq, chuncks=300):\n",
    "    return int(clamp(len(seq) / chuncks, 5, 10000))\n",
    "\n",
    "def rotate(p, theta):\n",
    "    theta = math.radians(theta)\n",
    "    xx = p[0]\n",
    "    yy = p[1]\n",
    "    xp = xx  * math.cos(theta) - yy*math.sin(theta)\n",
    "    yp = yy  * math.cos(theta) + xx*math.sin(theta)\n",
    "    return [xp, yp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3bd3f53-b9d6-4e4a-a086-43d0af1dd580",
   "metadata": {},
   "outputs": [],
   "source": [
    "dico_mau = {}\n",
    "with open(\"<dmau_scaffold2_V2.fasta>\") as fi:\n",
    "    for p, s in fasta_reader.get_seq_one_by_one(fi):\n",
    "        dico_mau[p] = s\n",
    "        \n",
    "dico_sim = {}\n",
    "with open(\"<dsim_scaffold2_V2.fasta>\") as fi:\n",
    "    for p, s in fasta_reader.get_seq_one_by_one(fi):\n",
    "        dico_sim[p] = s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a57fe3b7-1a15-44ef-bba5-61d06e341745",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = Path(\"<Figure_Dotplot_dir>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8eea889-0b8a-4468-b659-46232c92c7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# you need to get the CDS (exons) of the sequences \n",
    "# one sequence per file\n",
    "seq1_file = \"<dmau_cds_FBgn0267432.fna>\"\n",
    "seq2_file = \"<dsim_cds_FBgn0267432.fna>\"\n",
    "dico_s = {}\n",
    "with open(seq1_file) as fi:\n",
    "    for p, s in fasta_reader.get_seq_one_by_one(fi):\n",
    "        dico_s[\"mau\"] = s\n",
    "        mau_h = p\n",
    "with open(seq2_file) as fi:\n",
    "    for p, s in fasta_reader.get_seq_one_by_one(fi):\n",
    "        dico_s[\"sim\"] = s\n",
    "        sim_h = p\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeb237e1-5ee7-4a21-b13a-f128782d0421",
   "metadata": {},
   "outputs": [],
   "source": [
    "# then we nee to know the size of every exon.\n",
    "# The way I extract my CDS I got the information in the sequence name file (example: Y_scaffold1_FBgn0267432_FBtr0346771_kl-3__5910520...5911634;5913233...5913336;5913394...5914735;5970309...5970477;5970542...5970947;6178731...6178942;6178995...6179511;6203120...6203249;6203303...6205515;6205573...6206957;6270719...6271053;6298912...6299720;6299994...6301454;6343117...6345992;6346048...6346226;6486977...6487645__+)\n",
    "# here in the end you need a list with the size in pb of every exon in the same order of the exons\n",
    "\n",
    "mau_exon_size = []\n",
    "for r in mau_h.split(\"__\")[1].split(\";\"):\n",
    "    d = r.split('...')\n",
    "    mau_exon_size.append((int(d[1]) - int(d[0]) ))\n",
    "\n",
    "sim_exon_size = []\n",
    "for r in sim_h.split(\"__\")[1].split(\";\"):\n",
    "    d = r.split('...')\n",
    "    sim_exon_size.append((int(d[1]) - int(d[0]) ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a9b6658-8ced-47d2-990a-f55c14ed6fe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "windows_size = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e3ef78-08cb-423b-8aec-9de1f70e3ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_intron = np.array(sim_exon_size) / windows_size\n",
    "mau_intron = np.array(mau_exon_size) / windows_size\n",
    "\n",
    "# get the dot plot matrix for the coding sequence\n",
    "x = Rust_alnPairMat.get_matrix(dico_s[\"mau\"], dico_s[\"sim\"], windows_size, -10., -0.5, 5, True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919310cd-c4a6-4461-bfa4-0dfdad0712d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = \"kl-3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e4ecd0-e687-423a-83d5-6d5b65a74b1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure = plt.figure(figsize=(10, 10), dpi=80)\n",
    "ax = figure.add_axes((0,0,1,1))\n",
    "ax.imshow(x, vmin=0.5, cmap=\"Greys\", origin=\"lower\")\n",
    "plt.yticks(xticks, xlabels);\n",
    "plt.xticks(yticks, ylabels);\n",
    "xlim = ax.get_xlim()\n",
    "ylim = ax.get_ylim()\n",
    "\n",
    "\n",
    "s = 6\n",
    "p = 0\n",
    "patches = []\n",
    "for e in mau_intron[0:-1]:\n",
    "    p += e\n",
    "    p1 = rotate((0,0), -135)\n",
    "    p1[0] += p\n",
    "    p1[1] += p\n",
    "    \n",
    "    p2 = rotate((0.5*s,-1*s), -135)\n",
    "    p2[0] += p\n",
    "    p2[1] += p\n",
    "    \n",
    "    p3 = rotate((-0.5*s,-1*s), -135)\n",
    "    p3[0] += p\n",
    "    p3[1] += p\n",
    "    \n",
    "    polygon = Polygon((p1,\n",
    "                   p2,\n",
    "                   p3\n",
    "                  ),\n",
    "                    closed=True)     \n",
    "    patches.append(polygon)\n",
    "p = PatchCollection(patches, color=\"orange\")\n",
    "ax.add_collection(p)\n",
    "\n",
    "patches = []\n",
    "\n",
    "p = 0\n",
    "for e in sim_intron[0:-1]:\n",
    "    p += e\n",
    "    y_ = int(p)\n",
    "    p1 = rotate((0,0), 45)\n",
    "    p1[0] += p\n",
    "    p1[1] += p\n",
    "    \n",
    "    p2 = rotate((0.5*s,-1*s), 45)\n",
    "    p2[0] += p\n",
    "    p2[1] += p\n",
    "    \n",
    "    p3 = rotate((-0.5*s,-1*s), 45)\n",
    "    p3[0] += p\n",
    "    p3[1] += p\n",
    "    \n",
    "    polygon = Polygon((p1,\n",
    "                   p2,\n",
    "                   p3\n",
    "                  ),\n",
    "                    closed=True)     \n",
    "    patches.append(polygon)\n",
    "    #plt.vlines(p, 0, p, color=\"blue\")#, transform=ax.get_xaxis_transform())\n",
    "\n",
    "p = PatchCollection(patches, color=\"steelblue\")\n",
    "ax.add_collection(p)\n",
    " \n",
    "plt.ylabel(\"mau\")\n",
    "plt.xlabel(\"sim\")\n",
    "#plt.xlim(xlim)\n",
    "#plt.ylim(ylim)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "plt.title(\"cds {} dot plot\".format(gene_name))\n",
    "plt.savefig(out_dir / \"{}_cds.pdf\".format(gene_name), bbox_inches='tight')\n",
    "plt.savefig(out_dir / \"{}_cds.png\".format(gene_name), bbox_inches='tight')\n",
    "plt.show();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "328f1132-8dc4-4aca-8a0e-96892802773a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#intron need to do it manually foe every exon you want to:\n",
    "# get the sequences using the position. \n",
    "mau_i5 = dico_mau[\"Y_scaffold1\"][5970947-500  : 5970947 + 5000]\n",
    "sim_i5 = dico_sim[\"Y_scaffold2\"][220622-500  : 220622 + 5000]\n",
    "\n",
    "windows_size = 50\n",
    "x = Rust_alnPairMat.get_matrix(mau_i5, sim_i5, windows_size, -10., -0.5, 5, True)\n",
    "figure = plt.figure(figsize=(10, 10), dpi=80)\n",
    "ax = figure.add_axes((0,0,1,1))\n",
    "ax.imshow(x, vmin=0.5, cmap=\"Greys\", origin=\"lower\")\n",
    "plt.ylabel(\"mau\")\n",
    "plt.xlabel(\"sim\")\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "plt.savefig(out_dir / \"gene_intronN_start.pdf\", bbox_inches='tight')\n",
    "plt.show();\n",
    "plt.close();   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d29d789-206a-49a5-b89b-01a819ead8f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5763c64f-2285-409e-8226-3585d3934862",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
