{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "70bba0e6-c45d-4c31-9cd1-aaa77f237a07",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from Common import gtf_manipulation\n",
    "from pathlib import Path\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import sys\n",
    "import math\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23829828-3ddc-4874-bf5f-bff7e886cd59",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "51fb442d-b736-4e0b-b95f-e18c996352dd",
   "metadata": {},
   "source": [
    "## get intron exon gtf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10edd065-cf67-4b3c-abc5-d0cdeaedc2a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to extract exons and intron; \n",
    "# only gene with at least one intron are considered\n",
    "\n",
    "def get_intron(exons, gene_name):\n",
    "\n",
    "    if len(exons) <= 1:\n",
    "        return {}\n",
    "\n",
    "    intron = set()\n",
    "    exons = sorted(exons, key = lambda x: x[\"start\"])\n",
    "    for i in range(len(exons) - 1 ):\n",
    "        intron.add((exons[i][\"chr\"], exons[i][\"end\"]-1, exons[i+1][\"start\"]-1,  gene_name, \".\", exons[i][\"strand\"]))\n",
    "    return intron\n",
    "\n",
    "def get_exons(exons, gene_name):\n",
    "\n",
    "    if len(exons) <= 1:\n",
    "        return {}\n",
    "\n",
    "    intron = set()\n",
    "    exons = sorted(exons, key = lambda x: x[\"start\"])\n",
    "    for i in range(len(exons)):\n",
    "        intron.add((exons[i][\"chr\"], exons[i][\"start\"], exons[i][\"end\"],  gene_name, \".\", exons[i][\"strand\"]))\n",
    "    return intron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97f4b324-6eb5-4b7e-abd2-c3628a742268",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0fbd719-ac9f-405e-986f-28ed30eb77d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attr(string):\n",
    "    dico = {}\n",
    "    spt_ = string.split(\";\")\n",
    "    \n",
    "    for x in spt_:\n",
    "        if x:\n",
    "            try:\n",
    "                dico[x.split()[0].strip()] =  x.split()[1].replace('\"', \"\").strip()\n",
    "            except:\n",
    "                print(spt_)\n",
    "                raise\n",
    "    return dico\n",
    "\n",
    "\n",
    "def gtf_to_dict(gtf_file, main_ = \"gene_id\"):\n",
    "    print(main_)\n",
    "    dico = {}\n",
    "    with open(gtf_file) as f_in:\n",
    "        \"\"\"atm reads only genes\"\"\"\n",
    "        for line in f_in:\n",
    "            line = line.strip()\n",
    "            if not line:\n",
    "                continue\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            spt = line.strip().split(\"\\t\")\n",
    "            try:\n",
    "                chr_ = spt[0]\n",
    "                start = int(spt[3]) #- 1 # gtf are 1 based\n",
    "                end = int(spt[4]) #-1 \n",
    "                strand =  spt[6]\n",
    "                attr = get_attr(spt[-1])\n",
    "                \"\"\"                 if main_ not in attr:\n",
    "                                    print(attr)\n",
    "                continue \"\"\"\n",
    "                gene_id = attr[main_]\n",
    "                gene_symbol = attr.get(\"gene_symbol\", \"None\")\n",
    "                type_ = spt[2]\n",
    "            except:\n",
    "                print(attr)\n",
    "                print(line)\n",
    "                print(spt)\n",
    "                raise\n",
    "            \n",
    "            if gene_id not in dico:\n",
    "                dico[gene_id] = {\n",
    "                    \"chr\": chr_,\n",
    "                    \"symbol\" :  gene_symbol,\n",
    "                    \"strand\" : strand,\n",
    "                    \"transcript\": {}\n",
    "                }\n",
    "                \n",
    "            if type_ == \"gene\":\n",
    "    \n",
    "                dico[gene_id][\"start\"] = start\n",
    "                dico[gene_id][\"end\"] = end\n",
    "            \n",
    "            else:\n",
    "                try:\n",
    "                    transcript_id  = attr.get(\"transcript_id\", \"None\")\n",
    "                    transcript_symbol = attr.get(\"transcript_symbol\", \"None\")\n",
    "                    \n",
    "                    if transcript_id not in dico[gene_id][\"transcript\"]:\n",
    "                        dico[gene_id][\"transcript\"][transcript_id] = {\n",
    "                        \"transcript_symbol\" : transcript_symbol,\n",
    "                        \"transcript_id\" : transcript_id\n",
    "                        }\n",
    "                    \n",
    "                    if type_ not in dico[gene_id][\"transcript\"][transcript_id]: \n",
    "                        dico[gene_id][\"transcript\"][transcript_id][type_] = []\n",
    "                        \n",
    "                    dico[gene_id][\"transcript\"][transcript_id][type_].append({\n",
    "                        \"chr\": chr_,\n",
    "                        \"start\": start, \n",
    "                        \"end\" : end,\n",
    "                        \"strand\" : strand,\n",
    "                    })\n",
    "                except: \n",
    "                    continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a25e1928-ebc1-4a5a-8b78-9c9ebc1e0874",
   "metadata": {},
   "outputs": [],
   "source": [
    "dmau_gtfdict = gtf_manipulation.gtf_to_dict(\"Currated_Mau_genes_formatted.gtf\")\n",
    "dsim_gtfdict = gtf_manipulation.gtf_to_dict(\"Currated_Sim_genes_formatted.gtf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dfe6a8d-dfa5-4d6c-9819-f8edbe71fd9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sets = set()\n",
    "\n",
    "for k, v in dsim_gtfdict.items():\n",
    "    if tr_dico := v.get(\"transcript\"):\n",
    "        \n",
    "        for tr in tr_dico:\n",
    "            sets = sets.union(get_exons(tr_dico[tr][\"exon\"], \"{}_{}\".format(k, tr)))\n",
    "            \n",
    "            \n",
    "            \n",
    "with open(\"<dsim_exon.bed>\", \"w\") as fo:\n",
    "    seen = set()\n",
    "    for e in sets:\n",
    "        if (e[0], e[1], e[2]) in seen:\n",
    "            continue\n",
    "        else:\n",
    "            seen.add((e[0], e[1], e[2]))\n",
    "        fo.write(\"{}\\n\".format(\"\\t\".join(list(map(str, e)))))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0163426-f36a-4aef-979a-7b7217d89cf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sets = set()\n",
    "\n",
    "for k, v in dmau_gtfdict.items():\n",
    "    if tr_dico := v.get(\"transcript\"):\n",
    "        \n",
    "        for tr in tr_dico:\n",
    "            sets = sets.union(get_exons(tr_dico[tr][\"exon\"], \"{}_{}\".format(k, tr)))\n",
    "            \n",
    "            \n",
    "            \n",
    "with open(\"<dmau_exon.bed>\", \"w\") as fo:\n",
    "    seen = set()\n",
    "    for e in sets:\n",
    "        if (e[0], e[1], e[2]) in seen:\n",
    "            continue\n",
    "        else:\n",
    "            seen.add((e[0], e[1], e[2]))\n",
    "        fo.write(\"{}\\n\".format(\"\\t\".join(list(map(str, e)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1334b3-3c8f-4a85-b6c2-a9d3a4caad2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sets = set()\n",
    "\n",
    "for k, v in dsim_gtfdict.items():\n",
    "    if tr_dico := v.get(\"transcript\"):\n",
    "        \n",
    "        for tr in tr_dico:\n",
    "            sets = sets.union(get_intron(tr_dico[tr][\"exon\"], \"{}_{}\".format(k, tr)))\n",
    "            \n",
    "            \n",
    "            \n",
    "with open(\"<dsim_intron.bed>\", \"w\") as fo:\n",
    "    seen = set()\n",
    "    for e in sets:\n",
    "        if (e[0], e[1], e[2]) in seen:\n",
    "            continue\n",
    "        else:\n",
    "            seen.add((e[0], e[1], e[2]))\n",
    "        fo.write(\"{}\\n\".format(\"\\t\".join(list(map(str, e)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51aeea61-6c66-498c-88a4-dc29e30c34ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "sets = set()\n",
    "\n",
    "for k, v in dmau_gtfdict.items():\n",
    "    if tr_dico := v.get(\"transcript\"):\n",
    "        \n",
    "        for tr in tr_dico:\n",
    "            sets = sets.union(get_intron(tr_dico[tr][\"exon\"], \"{}_{}\".format(k, tr)))\n",
    "            \n",
    "            \n",
    "            \n",
    "with open(\"<dmau_intron.bed>\", \"w\") as fo:\n",
    "    seen = set()\n",
    "    for e in sets:\n",
    "        if (e[0], e[1], e[2]) in seen:\n",
    "            continue\n",
    "        else:\n",
    "            seen.add((e[0], e[1], e[2]))\n",
    "        fo.write(\"{}\\n\".format(\"\\t\".join(list(map(str, e)))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5187b2bf-5922-43ef-8c77-6357ced0d509",
   "metadata": {},
   "source": [
    "## match bed to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce263f4a-1201-42c8-a4f4-cce13e6544fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_bed_to_genome_file(bed_file, genome_file, out):\n",
    "    \n",
    "    dico_bed = {}\n",
    "    with open(bed_file) as fi:\n",
    "        for l in fi:\n",
    "            spt = l.strip().split()\n",
    "            chr_ = spt[0]\n",
    "            if chr_ not in dico_bed:\n",
    "                dico_bed[chr_] = []\n",
    "            dico_bed[chr_].append(l.strip())\n",
    "            \n",
    "    \n",
    "    with open(out, \"w\") as f_o, open(genome_file) as fi:\n",
    "        for l in fi:\n",
    "            spt = l.split()\n",
    "           \n",
    "            if spt[0] in dico_bed:\n",
    "                v = dico_bed[spt[0]]\n",
    "                if len(v) == 1:\n",
    "                    f_o.write(v[0] + \"\\n\")\n",
    "                else:\n",
    "                    f_o.write(\"\\n\".join(dico_bed[spt[0]]) + \"\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d501027-371d-4928-995b-24ef4ebce3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### you do have to replace with the actual path in your computer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85edb9ce-62f6-4e2b-a55c-4c5552aa8a76",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# run for the intron\n",
    "# for Dmau\n",
    "match_bed_to_genome_file(\"<sorted dmau intron bed file>\",\n",
    "                         \"<genome file dmau>\",\n",
    "                         \"<dmau_intron.bamsorted.bed>\")\n",
    "# for dsim\n",
    "match_bed_to_genome_file(\"<sorted dsim intron bed file>\",\n",
    "                         \"<genome file dsim>\",\n",
    "                         \"<dsim_intron.bamsorted.bed>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee939c90-0750-4857-8e88-4b5525065008",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run for the exon\n",
    "# for Dmau\n",
    "match_bed_to_genome_file(\"<sorted dmau exon bed file>\",\n",
    "                         \"<genome file dmau>\",\n",
    "                         \"<dmau_exon.bamsorted.bed>\")\n",
    "# for dsim\n",
    "match_bed_to_genome_file(\"<sorted dsim intron bed file>\",\n",
    "                         \"<genome file dsim>\",\n",
    "                         \"<dsim_exon.bamsorted.bed>\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e7b486-d7bc-444d-a67e-02c67c2b8966",
   "metadata": {},
   "source": [
    "## plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0174ff77-782c-470f-935b-f3effb9a11ea",
   "metadata": {},
   "source": [
    "### Exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc48b3fa-90d5-47e6-9d47-b536f8e74a77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# coverage file exon\n",
    "fsim = \"<coverage_dsimExon_dsim_SRR22548176>\"\n",
    "fmau = \"<coverage_dmauExon_dmau_SRR22548176>\"\n",
    "\n",
    "\n",
    "df_sim = pd.read_csv(fsim, sep=\"\\t\", header=None)\n",
    "df_mau = pd.read_csv(fmau, sep=\"\\t\", header=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fb1a00e-94ef-4b49-b6ff-1f3a4a6df026",
   "metadata": {},
   "source": [
    "### dmau"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "530f50ff-6449-4506-8fb3-a33e6c170f96",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig =  plt.figure(figsize=(10, 4))\n",
    "\n",
    "X = list(df_mau[~df_mau[3].str.contains(\"FBgn\")][8].values)\n",
    "Y = list(df_mau[~df_mau[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "Xp = list(df_mau[df_mau[3].str.contains(\"FBgn\")][8].values)\n",
    "Yp = list(df_mau[df_mau[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "ax_scat = fig.add_axes((0.0, 0, 0.60, 0.85))\n",
    "ax_hist_X_au = fig.add_axes((0.61, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "ax_hist_Y = fig.add_axes((0.80, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "\n",
    "\n",
    "ax_hist_X_au.hist(Y, color=\"#8da0cb\", orientation=\"horizontal\", range=(0, 1), bins=10)\n",
    "ax_hist_Y.hist(Yp, color=\"#fc8d62\", orientation=\"horizontal\", range=(0, 1), bins=10)\n",
    "\n",
    "\n",
    "#ax_hist.set_xlim(ax_hist.get_xlim()[::-1])\n",
    "ax_hist_X_au.set_yticks([])\n",
    "ax_hist_Y.set_yticks([])\n",
    "ax_scat.scatter(X, Y, marker=\".\", color=\"#8da0cb\", alpha=0.5)\n",
    "ax_scat.scatter(Xp, Yp, marker=\".\", color=\"#fc8d62\", alpha=1, s=100)\n",
    "\n",
    "\n",
    "ax_scat.set_yticks([0, 0.25, 0.5, 0.75, 1], labels=[\"0\", \"0.25\", \"0.5\", \"0.75\", \"1\"])\n",
    "#ax_scat.set_xticks([100_000, 400_000, 700_000], labels = [\"100kb\", \"400kb\", \"700kb\"])\n",
    "ax_scat.set(xlabel=\"exon len\", ylabel=\"ratio base covered\", title=\"exon bp covered ratio VS exon length\")\n",
    "#ax[1].set(xlabel=\"ratio coverage\", ylabel=\"intron len\")\n",
    "ax_hist_Y.tick_params(labelleft=False, left=False)\n",
    "ax_hist_X_au.tick_params(labelleft=False, left=False)\n",
    "\n",
    "ax_hist_Y.set(title = \"large Y genes\", xlabel = \"exon number\")\n",
    "ax_hist_X_au.set(title = \"X and Autosomes\", xlabel = \"exon number\")\n",
    "\n",
    "\n",
    "fig.suptitle('vs. D.mauritiana(Exon)')\n",
    "plt.savefig(\"dmau_exon_.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.savefig(\"dmau_exon.png\", bbox_inches=\"tight\")\n",
    "plt.show();\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d2880c3-6fdf-4190-af21-f3420d2efc1a",
   "metadata": {},
   "source": [
    "### D sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa7b8f87-0fd9-449d-9f65-7e0973ad5060",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig =  plt.figure(figsize=(10, 4))\n",
    "\n",
    "X = list(df_sim[~df_sim[3].str.contains(\"FBgn\")][8].values)\n",
    "Y = list(df_sim[~df_sim[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "Xp = list(df_sim[df_sim[3].str.contains(\"FBgn\")][8].values)\n",
    "Yp = list(df_sim[df_sim[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "ax_scat = fig.add_axes((0.0, 0, 0.60, 0.85))\n",
    "ax_hist_X_au = fig.add_axes((0.61, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "ax_hist_Y = fig.add_axes((0.80, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "\n",
    "\n",
    "ax_hist_X_au.hist(Y, color=\"#8da0cb\", orientation=\"horizontal\", range=(0, 1), bins=10)\n",
    "ax_hist_Y.hist(Yp, color=\"#fc8d62\", orientation=\"horizontal\", range=(0, 1), bins=10)\n",
    "\n",
    "\n",
    "#ax_hist.set_xlim(ax_hist.get_xlim()[::-1])\n",
    "ax_hist_X_au.set_yticks([])\n",
    "ax_hist_Y.set_yticks([])\n",
    "ax_scat.scatter(X, Y, marker=\".\", color=\"#8da0cb\", alpha=0.5)\n",
    "ax_scat.scatter(Xp, Yp, marker=\".\", color=\"#fc8d62\", alpha=1, s=100)\n",
    "\n",
    "\n",
    "ax_scat.set_yticks([0, 0.25, 0.5, 0.75, 1], labels=[\"0\", \"0.25\", \"0.5\", \"0.75\", \"1\"])\n",
    "#ax_scat.set_xticks([100_000, 400_000, 700_000], labels = [\"100kb\", \"400kb\", \"700kb\"])\n",
    "ax_scat.set(xlabel=\"exon len\", ylabel=\"ratio base covered\", title=\"exon bp covered ratio VS exon length\")\n",
    "#ax[1].set(xlabel=\"ratio coverage\", ylabel=\"intron len\")\n",
    "ax_hist_Y.tick_params(labelleft=False, left=False)\n",
    "ax_hist_X_au.tick_params(labelleft=False, left=False)\n",
    "\n",
    "ax_hist_Y.set(title = \"large Y genes\", xlabel = \"exon number\")\n",
    "ax_hist_X_au.set(title = \"X and Autosomes\", xlabel = \"exon number\")\n",
    "\n",
    "\n",
    "fig.suptitle('vs. D.simulans(exon)')\n",
    "plt.savefig(\"dsim_exon.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.savefig(\"dsim_exon.png\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa3ad6d5-b954-4950-8e81-d1e3720d9606",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Intron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "325d6f70-01db-4d05-b8c1-54fe7d903192",
   "metadata": {},
   "outputs": [],
   "source": [
    "# coverage file Intron\n",
    "fsim = \"<coverage_dsimIntron_dsim_SRR22548176>\"\n",
    "fmau = \"<coverage_dmauIntron_dmau_SRR22548176>\"\n",
    "\n",
    "\n",
    "df_sim = pd.read_csv(fsim, sep=\"\\t\", header=None)\n",
    "df_mau = pd.read_csv(fmau, sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "650784b4-27ab-4f19-8d1b-37dffd5cd22d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig =  plt.figure(figsize=(10, 4))\n",
    "\n",
    "X = list(df_mau[~df_mau[3].str.contains(\"FBgn\")][8].values)\n",
    "Y = list(df_mau[~df_mau[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "Xp = list(df_mau[df_mau[3].str.contains(\"FBgn\")][8].values)\n",
    "Yp = list(df_mau[df_mau[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "ax_scat = fig.add_axes((0.0, 0, 0.60, 0.85))\n",
    "ax_hist_X_au = fig.add_axes((0.61, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "ax_hist_Y = fig.add_axes((0.80, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "\n",
    "\n",
    "ax_hist_X_au.hist(Y, color=\"#8da0cb\", orientation=\"horizontal\", range=(0, 1), bins=15)\n",
    "ax_hist_Y.hist(Yp, color=\"#fc8d62\", orientation=\"horizontal\", range=(0, 1), bins=15)\n",
    "\n",
    "\n",
    "#ax_hist.set_xlim(ax_hist.get_xlim()[::-1])\n",
    "ax_hist_X_au.set_yticks([])\n",
    "ax_hist_Y.set_yticks([])\n",
    "ax_scat.scatter(X, Y, marker=\".\", color=\"#8da0cb\", alpha=0.5)\n",
    "ax_scat.scatter(Xp, Yp, marker=\".\", color=\"#fc8d62\", alpha=1, s=100)\n",
    "\n",
    "\n",
    "ax_scat.set_yticks([0, 0.25, 0.5, 0.75, 1], labels=[\"0\", \"0.25\", \"0.5\", \"0.75\", \"1\"])\n",
    "ax_scat.set_xticks([100_000, 400_000, 700_000], labels = [\"100kb\", \"400kb\", \"700kb\"])\n",
    "ax_scat.set(xlabel=\"intron len\", ylabel=\"ratio base covered\", title=\"intron bp covered ratio VS intron length\")\n",
    "#ax[1].set(xlabel=\"ratio coverage\", ylabel=\"intron len\")\n",
    "ax_hist_Y.tick_params(labelleft=False, left=False)\n",
    "ax_hist_X_au.tick_params(labelleft=False, left=False)\n",
    "\n",
    "ax_hist_Y.set(title = \"large Y genes\", xlabel = \"intron number\")\n",
    "ax_hist_X_au.set(title = \"X and Autosomes\", xlabel = \"intron number\")\n",
    "\n",
    "\n",
    "fig.suptitle('vs. D.mauritiana')\n",
    "plt.savefig(\"dmau_intron.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.savefig(\"dmau_intron.png\", bbox_inches=\"tight\")\n",
    "plt.show();\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79d8c9d1-a210-4c74-916a-deb00af06495",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig =  plt.figure(figsize=(10, 4))\n",
    "\n",
    "X = list(df_sim[~df_sim[3].str.contains(\"FBgn\")][8].values)\n",
    "Y = list(df_sim[~df_sim[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "Xp = list(df_sim[df_sim[3].str.contains(\"FBgn\")][8].values)\n",
    "Yp = list(df_sim[df_sim[3].str.contains(\"FBgn\")][9].values)\n",
    "\n",
    "ax_scat = fig.add_axes((0.0, 0, 0.60, 0.85))\n",
    "ax_hist_X_au = fig.add_axes((0.61, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "ax_hist_Y = fig.add_axes((0.80, 0, 0.18, 0.85), sharey=ax_scat, frameon=False)\n",
    "\n",
    "\n",
    "ax_hist_X_au.hist(Y, color=\"#8da0cb\", orientation=\"horizontal\", range=(0, 1), bins=10)\n",
    "ax_hist_Y.hist(Yp, color=\"#fc8d62\", orientation=\"horizontal\", range=(0, 1), bins=10)\n",
    "\n",
    "#ax_hist.set_xlim(ax_hist.get_xlim()[::-1])\n",
    "ax_hist_X_au.set_yticks([])\n",
    "ax_hist_Y.set_yticks([])\n",
    "ax_scat.scatter(X, Y, marker=\".\", color=\"#8da0cb\", alpha=0.5)\n",
    "ax_scat.scatter(Xp, Yp, marker=\".\", color=\"#fc8d62\", alpha=1, s=100)\n",
    "\n",
    "\n",
    "ax_scat.set_yticks([0, 0.25, 0.5, 0.75, 1], labels=[\"0\", \"0.25\", \"0.5\", \"0.75\", \"1\"])\n",
    "ax_scat.set_xticks([100_000, 400_000, 700_000], labels = [\"100kb\", \"400kb\", \"700kb\"])\n",
    "ax_scat.set(xlabel=\"intron len\", ylabel=\"ratio base covered\", title=\"intron bp covered ratio VS intron length\")\n",
    "#ax[1].set(xlabel=\"ratio coverage\", ylabel=\"intron len\")\n",
    "ax_hist_Y.tick_params(labelleft=False, left=False)\n",
    "ax_hist_X_au.tick_params(labelleft=False, left=False)\n",
    "\n",
    "ax_hist_Y.set(title = \"large Y genes\", xlabel = \"intron number\")\n",
    "ax_hist_X_au.set(title = \"X and Autosomes\", xlabel = \"intron number\")\n",
    "\n",
    "\n",
    "fig.suptitle('vs. D.simulans')\n",
    "plt.savefig(\"dsim_intron.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.savefig(\"dsim_intron.png\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b65d09fc-e617-4ef8-8a75-8a8c141a584c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "714ee43a-e630-4445-8ef3-8c5495afebd1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "805d0ac7-fbb8-43f3-ade2-8dc062ad803e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
