{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65dbf690-3cbc-4179-8152-3a76f8e5c61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from pathlib import Path\n",
    "import re\n",
    "import math\n",
    "import numpy as np\n",
    "from matplotlib.lines import Line2D\n",
    "import matplotlib.patches as mpatches\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe6e6cd-e91a-4e88-a596-c257a75d60c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attr(string):\n",
    "    dico = {}\n",
    "    spt_ = string.split(\";\")\n",
    "    \n",
    "    for x in spt_:\n",
    "        if x:\n",
    "            try:\n",
    "                dico[x.split()[0].strip()] =  x.split()[1].replace('\"', \"\").strip()\n",
    "            except:\n",
    "                print(spt_)\n",
    "                raise\n",
    "    return dico\n",
    "\n",
    "\n",
    "def gtf_to_dict(gtf_file, main_ = \"gene_id\"):\n",
    "    print(main_)\n",
    "    dico = {}\n",
    "    with open(gtf_file) as f_in:\n",
    "        \"\"\"atm reads only genes\"\"\"\n",
    "        for line in f_in:\n",
    "            line = line.strip()\n",
    "            if not line:\n",
    "                continue\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            spt = line.strip().split(\"\\t\")\n",
    "            try:\n",
    "                chr_ = spt[0]\n",
    "                start = int(spt[3]) #- 1 # gtf are 1 based\n",
    "                end = int(spt[4]) #-1 \n",
    "                strand =  spt[6]\n",
    "                attr = get_attr(spt[-1])\n",
    "                \"\"\"                 if main_ not in attr:\n",
    "                                    print(attr)\n",
    "                continue \"\"\"\n",
    "                gene_id = attr[main_]\n",
    "                gene_symbol = attr.get(\"gene_symbol\", \"None\")\n",
    "                type_ = spt[2]\n",
    "            except:\n",
    "                print(attr)\n",
    "                print(line)\n",
    "                print(spt)\n",
    "                raise\n",
    "            \n",
    "            if gene_id not in dico:\n",
    "                dico[gene_id] = {\n",
    "                    \"chr\": chr_,\n",
    "                    \"symbol\" :  gene_symbol,\n",
    "                    \"strand\" : strand,\n",
    "                    \"transcript\": {}\n",
    "                }\n",
    "                \n",
    "            if type_ == \"gene\":\n",
    "    \n",
    "                dico[gene_id][\"start\"] = start\n",
    "                dico[gene_id][\"end\"] = end\n",
    "            \n",
    "            else:\n",
    "                try:\n",
    "                    transcript_id  = attr.get(\"transcript_id\", \"None\")\n",
    "                    transcript_symbol = attr.get(\"transcript_symbol\", \"None\")\n",
    "\n",
    "                    # gene_id  = attr[\"gene_id\"]\n",
    "                    # gene_symbol = attr.get(\"gene_symbol\", \"None\")\n",
    "                    # strand =  spt[6]\n",
    "                    \n",
    "                    if transcript_id not in dico[gene_id][\"transcript\"]:\n",
    "                        dico[gene_id][\"transcript\"][transcript_id] = {\n",
    "                        \"transcript_symbol\" : transcript_symbol,\n",
    "                        \"transcript_id\" : transcript_id\n",
    "                        }\n",
    "                    \n",
    "                    if type_ not in dico[gene_id][\"transcript\"][transcript_id]: \n",
    "                        dico[gene_id][\"transcript\"][transcript_id][type_] = []\n",
    "                        \n",
    "                    dico[gene_id][\"transcript\"][transcript_id][type_].append({\n",
    "                        \"chr\": chr_,\n",
    "                        \"start\": start, \n",
    "                        \"end\" : end,\n",
    "                        \"strand\" : strand,\n",
    "                    })\n",
    "                except: \n",
    "                    continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34da552e-2481-482b-afe6-8180a18e7d38",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = \"<path to annotated Y genes gtf>\"\n",
    "gtf_dict = gtf_to_dict(gtf_file, \"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1cb0bad-9d7c-4549-9281-54f77ff4d691",
   "metadata": {},
   "outputs": [],
   "source": [
    "omnisplice_run = Path(\"<Path to directory containing OmniSplice output>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a118858-c598-4b27-a018-ecef9c85eabb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def data_to_df_for_glm(control, treatment):\n",
    "    df = pd.DataFrame(control + treatment)\n",
    "    df[\"failures\"] = df.apply(lambda x: sum(x[1:]), axis=1)\n",
    "    df[\"successes\"] = df[0]\n",
    "    df[\"group\"] = [\"control\"] * len(control) + [\"treatment\"] * len(treatment)\n",
    "    return df\n",
    "\n",
    "def gln_binomial_test(control, treatment):\n",
    "    df = data_to_df_for_glm(control, treatment)\n",
    "    mod = smf.glm('successes + failures ~ group', family=sm.families.Binomial(), data=df).fit()\n",
    "    # convert intercept to probability\n",
    "    odds_i = math.exp(mod.params[0])\n",
    "    control_p = odds_i / (1 + odds_i)\n",
    "\n",
    "    # convert stable=\"stable\" to probability\n",
    "    odds_p = math.exp( mod.params[0]) * math.exp(mod.params[1])\n",
    "    treatment_p = (odds_p / (1 + odds_p))\n",
    "    return (mod.pvalues[1], control_p , treatment_p)\n",
    "\n",
    "\n",
    "def collapse(values):\n",
    "    res = []\n",
    "    for i, v in enumerate(values):\n",
    "        if i % 2 == 0:\n",
    "            res.append([x+values[i+1][ii] for ii, x in enumerate(v)])\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a3ebb6-6ee6-4ee6-b398-8f9876f3d5f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = re.compile(\"(.*)_\\d+_S\\d\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d98702c-ee75-4707-9915-d3e3e15caf55",
   "metadata": {},
   "outputs": [],
   "source": [
    "dico_result = {}\n",
    "# merged\n",
    "for file in sorted(omnisplice_run.glob(\"*table\")):\n",
    "    print(file)\n",
    "    geno = reg.search(file.stem).group(1) \n",
    "\n",
    "    \n",
    "    with open(file) as fi:\n",
    "        header = fi.readline().strip()\n",
    "        header = dict((v, i) for i,v in enumerate(header.split()))\n",
    "        for l in fi:\n",
    "            spt = l.strip().split('\\t')\n",
    "            value = []\n",
    "            value = list(map(int, spt[header[\"spliced\"]: header[\"exon_intron\"] + 1]))\n",
    "            value[-1] += int(spt[header[\"exon_other\"]])\n",
    "\n",
    "            gene = spt[1]\n",
    "            if gene not in dico_result:\n",
    "                dico_result[gene] = {}\n",
    "            if spt[header[\"exon_type\"]] == \"Acceptor\":\n",
    "                continue\n",
    "            exon_n = spt[3]\n",
    "            if exon_n not in dico_result[gene] :\n",
    "                dico_result[gene][exon_n] = {}\n",
    "            \n",
    "            if geno not in dico_result[gene][exon_n]:\n",
    "                dico_result[gene][exon_n][geno] = []\n",
    "            dico_result[gene][exon_n][geno].append(value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41da8e3e-08fc-442b-a908-ad7ad00f187a",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = Path(\"<path to output dir>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71f3b4bb-deae-4585-9f77-99a8dec5745b",
   "metadata": {},
   "outputs": [],
   "source": [
    "color = [\"#D3D3D385\", \"#214d4e\",  \"#cc655b\", '#41bbc5', \"#c6dbae\", '#069668',  '#b3e61c',  '#d55e00', '#cc78bc',\n",
    "         '#ca9161', '#fbafe4',  '#029e73']\n",
    "mpl.rcParams['axes.linewidth'] = 1.5\n",
    "mpl.rcParams[\"axes.spines.right\"] = False\n",
    "mpl.rcParams[\"axes.spines.top\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dfd9bdf-57d9-4495-a1f0-5b52570c9aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for gene in dico_result:\n",
    "    try:\n",
    "        gene_name = dico_gtf[gene][\"symbol\"]\n",
    "    except:\n",
    "        print(gene_name)\n",
    "        gene_name = 'PRY'\n",
    "    if gene_name == \"PRY_alt\":\n",
    "        gene_name = \"PRY\"\n",
    "        \n",
    "    len_exon = len(dico_result[gene])\n",
    "    z = 0\n",
    "    exon_order = list(sorted(dico_result[gene].keys(), key=lambda x: x.split('_')[0], reverse=True))[:-1]\n",
    "    for i_exon, exon in enumerate(exon_order):\n",
    "           \n",
    "        pure = dico_result[gene][exon][\"mau\"]\n",
    "        control = pure\n",
    "        \n",
    "        pure = np.sum(np.array(pure), axis=0)\n",
    "        s_pure = sum(pure) \n",
    "        pure = pure / s_pure\n",
    "        hyb = dico_result[gene][exon][\"sim_mau\"]\n",
    "        treatment = hyb\n",
    "        hyb = np.sum(np.array(hyb), axis=0)\n",
    "        s_hyb = sum(hyb) \n",
    "        hyb = hyb / s_hyb\n",
    "        this = exon_order[i_exon]\n",
    "        this = this.split('_')\n",
    "        this[1] = str(int(this[1]))\n",
    "        exon_order[i_exon] = \" \".join(this)\n",
    "        print(pure, hyb)\n",
    "        try:\n",
    "            pval, control_p, treat_p = gln_binomial_test(control, treatment)\n",
    "            if pval < 0.001:\n",
    "                \n",
    "                exon_order[i_exon] += \" *\"\n",
    "            print(i_exon, pval)    \n",
    "        except:\n",
    "            print(control, treatment)\n",
    "            exon_order[i_exon] += \"_na\"\n",
    "        \n",
    "        bottom = 0\n",
    "        for i,e in enumerate(pure):\n",
    "            plt.barh(y=z, width=e,left=bottom, color = color[i])\n",
    "            bottom += e\n",
    "        bottom += 0.05\n",
    "        for i,e in enumerate(hyb[::-1]):\n",
    "            plt.barh(y=z, width=e,left=bottom, color = color[3-i])\n",
    "            bottom += e\n",
    "    \n",
    "        z += 1\n",
    "\n",
    "    plt.yticks(ticks=range(0,z), labels=exon_order)\n",
    "    handles, labels = plt.gca().get_legend_handles_labels()\n",
    "    line1 = Line2D([], [], label=\"Spliced\",   color=color[0], linewidth=6)\n",
    "    line2 = Line2D([], [], label=\"Unspliced\",  color=color[1], linewidth=6)\n",
    "    line3 = Line2D([], [], label=\"Clipped\",   color=color[2], linewidth=6)\n",
    "    line4 = Line2D([], [], label=\"Exon_Intron\",   color=color[3], linewidth=6)\n",
    "    \n",
    "    handles.extend([ line1, line2, line3, line4])#, line5])\n",
    "    plt.gcf().legend(handles=handles, loc='outside right upper', bbox_to_anchor=(1.25, 0.8))\n",
    "    plt.gca().set_xticks([])\n",
    "    plt.title(gene_name,     fontsize=24)\n",
    "    plt.xlim(-0.1, 2.2)\n",
    "\n",
    "    plt.tick_params(axis='y', labelsize=12)\n",
    "\n",
    "    plt.savefig(out_dir / \"{}.pdf\".format(gene_name), bbox_inches=\"tight\")\n",
    "    plt.show();\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
